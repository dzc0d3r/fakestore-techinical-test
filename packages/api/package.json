{
  "name": "api",
  "version": "1.0.0",
  "description": "",
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    }
  },
  "scripts": {
    "build": "tsup src/index.ts --format cjs,esm --dts",
    "dev": "tsup src/index.ts --format cjs,esm --dts --watch",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "packageManager": "pnpm@10.7.1",
  "dependencies": {
    "@tanstack/react-query": "^5.71.10",
    "axios": "^1.8.4"
  },
  "devDependencies": {
    "@types/jsonwebtoken": "^9.0.9",
    "tsup": "^8.4.0",
    "typescript": "^5.8.3"
  }
}
