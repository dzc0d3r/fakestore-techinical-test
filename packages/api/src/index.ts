import axios from "axios";
import {
  useProducts,
  useProduct,
  useCreateProduct,
  useUpdateProduct,
  useDeleteProduct
} from "@/hooks/use-products"
import type { Product } from "@/types/products";

export type LoginCredentials = {
  username: string;
  password: string;
};

export type LoginResponse = {
  token?: string;
  error?: any;
};

export const api = axios.create({
  baseURL: "https://fakestoreapi.com",
});

export const login = async (credentials: LoginCredentials): Promise<LoginResponse> => {
  try {
    const response = await api.post<LoginResponse>('auth/login', credentials);
    return response.data;
  } catch (error) {
    return error as any;
  }
};

export {
  useProducts,
  useProduct,
  useCreateProduct,
  useUpdateProduct,
  useDeleteProduct
};
export type { Product };
