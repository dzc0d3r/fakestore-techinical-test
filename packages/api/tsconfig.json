{
  "compilerOptions": {
    "target": "ES2020",
    "module": "NodeNext",
    "lib": ["ES2020", "DOM"],
    "moduleResolution": "NodeNext",
    "declaration": true,
    "outDir": "./dist",
    "strict": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "jsx": "react-jsx",
    "paths": {
      "@/*": ["./src/*"],
    }
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules"]
}
