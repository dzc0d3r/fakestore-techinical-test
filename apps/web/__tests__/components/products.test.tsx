import Products from "@/components/products";
import { render } from "@testing-library/react";
import { useProducts } from "api";
import { describe, expect, it, vi } from "vitest";

// Mock next-auth session provider and hooks
vi.mock("next-auth/react", () => ({
  useSession: () => ({ data: null, status: "unauthenticated" }),
  SessionProvider: ({ children }: { children: React.ReactNode }) => children,
}));

// Mock API module
vi.mock("api", () => ({
  useProducts: vi.fn(),
}));

// Mock child components that might use session
vi.mock("./single-product", () => ({
  default: ({ product }: any) => (
    <div data-testid="single-product">{product.name}</div>
  ),
}));

// Mock cart component if it uses session
vi.mock("./cart", () => ({
  default: () => <div data-testid="cart" />,
}));

// Mock skeleton component
vi.mock("@/components/ui/skeleton", () => ({
  Skeleton: ({ className }: { className?: string }) => (
    <div className={className} data-testid="skeleton" />
  ),
}));

describe("Products", () => {
  const mockUseProducts = useProducts as vi.Mock;

  it("renders loading state correctly", () => {
    mockUseProducts.mockReturnValue({
      data: null,
      isLoading: true,
      error: null,
    });

    const { container } = render(<Products />);
    expect(container).toMatchSnapshot();
  });

  it("renders error state correctly", () => {
    mockUseProducts.mockReturnValue({
      data: null,
      isLoading: false,
      error: { message: "Failed to load products" },
    });

    const { container } = render(<Products />);
    expect(container).toMatchSnapshot();
  });

  it("renders products correctly", () => {
    const mockProducts = [
      { id: 1, title: "Product 1", price: 100 },
      { id: 2, title: "Product 2", price: 200 },
    ];

    mockUseProducts.mockReturnValue({
      data: mockProducts,
      isLoading: false,
      error: null,
    });

    const { container } = render(<Products />);
    expect(container).toMatchSnapshot();
  });
});
