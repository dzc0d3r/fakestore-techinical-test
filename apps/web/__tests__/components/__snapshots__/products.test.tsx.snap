// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Products > renders error state correctly 1`] = `
<div>
  <div
    class="p-4 text-red-500"
  >
    Error: 
    Failed to load products
  </div>
</div>
`;

exports[`Products > renders loading state correctly 1`] = `
<div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative group"
  >
    <div
      class="h-96 flex flex-col p-5 rounded hover:scale-105 transition-transform ease-out duration-200"
    >
      <div
        class="absolute top-0 z-10"
      >
        <div
          class="h-8 w-8 rounded-full"
          data-testid="skeleton"
        />
      </div>
      <div
        class="relative h-72 max-h-72 mb-4"
      >
        <div
          class="h-full w-full rounded-lg"
          data-testid="skeleton"
        />
      </div>
      <div
        class="font-semibold flex items-center justify-between mt-4 mb-1"
      >
        <div
          class="h-4 w-32"
          data-testid="skeleton"
        />
        <div
          class="h-4 w-12"
          data-testid="skeleton"
        />
      </div>
      <div
        class="space-y-1"
      >
        <div
          class="h-3 w-full"
          data-testid="skeleton"
        />
        <div
          class="h-3 w-4/5"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Products > renders products correctly 1`] = `
<div>
  <div
    class="flex flex-col relative"
  >
    <div
      class="h-96 flex flex-col p-5 group rounded hover:scale-105 transition-transform ease-out duration-200 "
    >
      <div
        class="absolute top-0 z-10"
      />
      <a
        href="/products/1"
      >
        <div
          class="relative h-72 max-h-72"
        >
          <img
            alt="Product 1"
            class="absolute z-auto object-contain duration-700 ease-in-out group-hover:opacity-75 scale-100 blur-2xl grayscale}"
            data-nimg="fill"
            decoding="async"
            src=""
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
        <div
          class="font-semibold flex items-center justify-between mt-4 mb-1"
        >
          <p
            class="w-44 truncate"
          >
            Product 1
          </p>
          <p>
            $
            100
          </p>
        </div>
        <p
          class="italic text-xs mx-w-58 line-clamp-2 text-gray-600"
        />
      </a>
    </div>
  </div>
  <div
    class="flex flex-col relative"
  >
    <div
      class="h-96 flex flex-col p-5 group rounded hover:scale-105 transition-transform ease-out duration-200 "
    >
      <div
        class="absolute top-0 z-10"
      />
      <a
        href="/products/2"
      >
        <div
          class="relative h-72 max-h-72"
        >
          <img
            alt="Product 2"
            class="absolute z-auto object-contain duration-700 ease-in-out group-hover:opacity-75 scale-100 blur-2xl grayscale}"
            data-nimg="fill"
            decoding="async"
            src=""
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
        <div
          class="font-semibold flex items-center justify-between mt-4 mb-1"
        >
          <p
            class="w-44 truncate"
          >
            Product 2
          </p>
          <p>
            $
            200
          </p>
        </div>
        <p
          class="italic text-xs mx-w-58 line-clamp-2 text-gray-600"
        />
      </a>
    </div>
  </div>
</div>
`;
